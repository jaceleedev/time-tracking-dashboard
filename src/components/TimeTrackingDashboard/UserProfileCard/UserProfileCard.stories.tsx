import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import UserProfileCard from './UserProfileCard';
import { useState } from 'react';

const meta = {
  component: UserProfileCard,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    activeTimeframe: {
      description: 'Currently active timeframe',
      control: 'select',
      options: ['daily', 'weekly', 'monthly'],
    },
    onTimeframeChange: {
      description: 'Function to call when timeframe is changed',
    },
  },
  args: {
    onTimeframeChange: fn(),
  },
  decorators: [
    (Story) => (
      <div style={{ width: '327px' }}>
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof UserProfileCard>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    activeTimeframe: 'weekly',
  },
};

export const Active: Story = {
  render: () => {
    const [activeTimeframe, setActiveTimeframe] = useState<
      'daily' | 'weekly' | 'monthly'
    >('weekly');
    return (
      <UserProfileCard
        activeTimeframe={activeTimeframe}
        onTimeframeChange={setActiveTimeframe}
      />
    );
  },
  args: {
    activeTimeframe: 'weekly',
  },
};
