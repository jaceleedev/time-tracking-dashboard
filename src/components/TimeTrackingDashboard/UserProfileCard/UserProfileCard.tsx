import Image from 'next/image';
import { TimeframeOption } from '@/types/dashboard';

interface UserProfileCardProps {
  activeTimeframe: TimeframeOption;
  onTimeframeChange: (timeframe: TimeframeOption) => void;
}

function UserProfileCard({
  activeTimeframe,
  onTimeframeChange,
}: Readonly<UserProfileCardProps>) {
  return (
    <article className="w-full h-[203px] rounded-[15px] bg-dark-blue sm:col-span-2 sm:w-3/5 sm:mx-auto lg:col-span-1 lg:w-full lg:h-[518px] lg:row-span-2">
      <h2 className="sr-only">User Profile Card</h2>
      <div className="flex justify-center items-center h-[133px] rounded-[15px] bg-blue lg:justify-start lg:items-start lg:h-[354px] lg:pl-8 lg:pt-[37px]">
        <div className="flex items-center gap-5 lg:flex-col lg:items-start  lg:gap-[43px]">
          <Image
            src="/images/image-jeremy.png"
            alt="Jeremy Robson's profile picture"
            width={64}
            height={64}
            priority
            className="border-[3px] border-solid border-white rounded-full lg:w-[78px] lg:h-[78px]"
          />
          <div className="flex flex-col justify-center items-start h-[50px] pr-[1px] gap-y-1 lg:justify-start lg:w-[138px] lg:h-[116px] lg:gap-y-[3px] ">
            <p className="text-report-title text-pale-blue">Report for</p>
            <p className="text-user-name text-white">Jeremy Robson</p>
          </div>
        </div>
      </div>
      <nav
        className="flex items-center h-[70px] lg:flex-col lg:items-start lg:gap-y-[21px] lg:h-[164px] lg:pt-[26px] lg:pl-8 lg:pb-8"
        aria-label="Timeframe selection"
      >
        {(['daily', 'weekly', 'monthly'] as const).map((timeframe) => (
          <button
            key={timeframe}
            onClick={() => onTimeframeChange(timeframe)}
            className={`basis-1/3 text-tab hover:text-white ${activeTimeframe === timeframe ? 'text-white' : 'text-desaturated-blue'}`}
            aria-current={activeTimeframe === timeframe ? 'true' : 'false'}
            aria-label={`View ${timeframe} Report`}
          >
            {timeframe.charAt(0).toUpperCase() + timeframe.slice(1)}
          </button>
        ))}
      </nav>
    </article>
  );
}

export default UserProfileCard;
