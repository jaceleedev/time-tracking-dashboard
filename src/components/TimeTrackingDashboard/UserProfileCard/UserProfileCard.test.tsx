import { describe, afterEach, expect, it, vi } from 'vitest';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import UserProfileCard from './UserProfileCard';

expect.extend(toHaveNoViolations);

describe('UserProfileCard', () => {
  const mockOnTimeframeChange = vi.fn();

  afterEach(() => {
    vi.clearAllMocks();
    cleanup();
  });

  // 1. Rendering test
  it('renders correctly', () => {
    render(
      <UserProfileCard
        activeTimeframe="weekly"
        onTimeframeChange={mockOnTimeframeChange}
      />
    );

    const userName = screen.getByText('Jeremy Robson');
    const reportFor = screen.getByText('Report for');
    const dailyButton = screen.getByText('Daily');
    const weeklyButton = screen.getByText('Weekly');
    const monthlyButton = screen.getByText('Monthly');

    expect(userName).toBeDefined();
    expect(reportFor).toBeDefined();
    expect(dailyButton).toBeDefined();
    expect(weeklyButton).toBeDefined();
    expect(monthlyButton).toBeDefined();
  });

  // 2. Props test
  it('applies correct styles based on activeTimeframe prop', () => {
    render(
      <UserProfileCard
        activeTimeframe="weekly"
        onTimeframeChange={mockOnTimeframeChange}
      />
    );

    const weeklyButton = screen.getByText('Weekly');
    const dailyButton = screen.getByText('Daily');

    expect(weeklyButton.className).toContain('text-white');
    expect(weeklyButton.getAttribute('aria-current')).toBe('true');
    expect(dailyButton.className).toContain('text-desaturated-blue');
    expect(dailyButton.getAttribute('aria-current')).toBe('false');
  });

  // 3. Styles test
  it('renders the profile image with correct styles', () => {
    render(
      <UserProfileCard
        activeTimeframe="weekly"
        onTimeframeChange={mockOnTimeframeChange}
      />
    );

    const profileImage = screen.getByAltText("Jeremy Robson's profile picture");

    expect(profileImage).toBeDefined();
    expect(profileImage).toHaveProperty('src');
    expect(profileImage.className).toContain(
      'border-[3px] border-solid border-white rounded-full'
    );
  });

  // 4. Behavior test
  it('calls onTimeframeChange when a timeframe button is clicked', () => {
    render(
      <UserProfileCard
        activeTimeframe="weekly"
        onTimeframeChange={mockOnTimeframeChange}
      />
    );

    const dailyButton = screen.getByText('Daily');
    fireEvent.click(dailyButton);
    expect(mockOnTimeframeChange).toHaveBeenCalledWith('daily');

    const monthlyButton = screen.getByText('Monthly');
    fireEvent.click(monthlyButton);
    expect(mockOnTimeframeChange).toHaveBeenCalledWith('monthly');
  });

  // 5. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <UserProfileCard
        activeTimeframe="weekly"
        onTimeframeChange={mockOnTimeframeChange}
      />
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  // 6. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <UserProfileCard
        activeTimeframe="weekly"
        onTimeframeChange={mockOnTimeframeChange}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
