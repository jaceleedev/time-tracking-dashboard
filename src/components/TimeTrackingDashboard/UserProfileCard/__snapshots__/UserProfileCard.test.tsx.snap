// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserProfileCard > matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="w-full h-[203px] rounded-[15px] bg-dark-blue sm:col-span-2 sm:w-3/5 sm:mx-auto lg:col-span-1 lg:w-full lg:h-[518px] lg:row-span-2"
  >
    <h2
      class="sr-only"
    >
      User Profile Card
    </h2>
    <div
      class="flex justify-center items-center h-[133px] rounded-[15px] bg-blue lg:justify-start lg:items-start lg:h-[354px] lg:pl-8 lg:pt-[37px]"
    >
      <div
        class="flex items-center gap-5 lg:flex-col lg:items-start  lg:gap-[43px]"
      >
        <img
          alt="Jeremy Robson's profile picture"
          class="border-[3px] border-solid border-white rounded-full lg:w-[78px] lg:h-[78px]"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="64"
          src="/_next/image?url=%2Fimages%2Fimage-jeremy.png&w=128&q=75"
          srcset="/_next/image?url=%2Fimages%2Fimage-jeremy.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fimage-jeremy.png&w=128&q=75 2x"
          style="color: transparent;"
          width="64"
        />
        <div
          class="flex flex-col justify-center items-start h-[50px] pr-[1px] gap-y-1 lg:justify-start lg:w-[138px] lg:h-[116px] lg:gap-y-[3px] "
        >
          <p
            class="text-report-title text-pale-blue"
          >
            Report for
          </p>
          <p
            class="text-user-name text-white"
          >
            Jeremy Robson
          </p>
        </div>
      </div>
    </div>
    <nav
      aria-label="Timeframe selection"
      class="flex items-center h-[70px] lg:flex-col lg:items-start lg:gap-y-[21px] lg:h-[164px] lg:pt-[26px] lg:pl-8 lg:pb-8"
    >
      <button
        aria-current="false"
        aria-label="View daily Report"
        class="basis-1/3 text-tab hover:text-white text-desaturated-blue"
      >
        Daily
      </button>
      <button
        aria-current="true"
        aria-label="View weekly Report"
        class="basis-1/3 text-tab hover:text-white text-white"
      >
        Weekly
      </button>
      <button
        aria-current="false"
        aria-label="View monthly Report"
        class="basis-1/3 text-tab hover:text-white text-desaturated-blue"
      >
        Monthly
      </button>
    </nav>
  </article>
</DocumentFragment>
`;
